### PR TITLE
[ROCm] enable hipGraph

### PR DESCRIPTION
**Pull request by jeffdaily**
This ports the cudaGraph support from the CUDA EP to the ROCM EP's hipGraph.

### CODE REVIEW

**Comment by tianleiwu** [comment]
Update the CUDA Graph unit test to cover hipGraph?

File: onnxruntime/core/session/inference_session.cc [hunk 1]
```diff
@@ -144,7 +144,7 @@ static bool AreAllComputeNodesAssignedToCudaEp(const Graph& graph) {
 
     // Empty node provider means CPU EP
     if (!node_provider.empty() &&
-        node_provider != kCudaExecutionProvider &&
+        !(node_provider == kCudaExecutionProvider || node_provider == kRocmExecutionProvider) &&
         node_provider != kCpuExecutionProvider) {
       nodes_on_cpu_and_cuda_eps_only = false;
```
Comment (hariharans29) [inline onnxruntime/core/session/inference_session.cc:149]: Nit: Var name to reflect rocm ?

File: onnxruntime/core/providers/rocm/rocm_execution_provider.cc [hunk 2]
```diff
@@ -170,13 +170,42 @@ ROCMExecutionProvider::PerThreadContext::PerThreadContext(OrtDevice::DeviceId de
...
+  hip_graph_.CaptureEnd();
+  is_graph_captured_ = true;
+}
+
+bool ROCMExecutionProvider::PerThreadContext::IsGraphCaptured() const {
+  return is_graph_captured_;
+}
+
+Status ROCMExecutionProvider::PerThreadContext::ReplayGraph() {
+  ORT_ENFORCE(IsGraphCaptured());
```
Comment (cloudhan) [inline onnxruntime/core/providers/rocm/rocm_execution_provider.cc:201]: This function returns a Status, you could switch to use return an Status to indicate the Error. Better also and some message about thoes enforcement.

**Comment by cloudhan** [comment]
@hariharans29  @tianleiwu  Do we have any test around CUDA graph? If yes, maybe some `#ifdef`s should be taken care.

File: onnxruntime/core/providers/rocm/rocm_execution_provider.cc (hunk 2 above)
Comment (jeffdaily) [reply onnxruntime/core/providers/rocm/rocm_execution_provider.cc:201]: This function was ported from the CUDA EP, so perhaps CUDA EP needs a fix too?

File: onnxruntime/core/session/inference_session.cc (hunk 1 above)
Comment (jeffdaily) [reply onnxruntime/core/session/inference_session.cc:149]: Sorry I don't understand your nit?  The new line has kRocmExecutionProvider to reflect rocm?

**Comment by jeffdaily** [comment]
> Update the CUDA Graph unit test to cover hipGraph?

Is this the only one?

onnxruntime/onnxruntime/test/python/onnxruntime_test_python_cudagraph.py

**Comment by tianleiwu** [comment]
> > Update the CUDA Graph unit test to cover hipGraph?
> 
> Is this the only one?
> 
> onnxruntime/onnxruntime/test/python/onnxruntime_test_python_cudagraph.py
Please make sure this test can cover hipGraph.

File: onnxruntime/core/session/inference_session.cc (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/core/session/inference_session.cc:149]: Like `nodes_on_cpu_cuda_rocm_eps_only`

**Comment by cloudhan** [comment]
@jeffdaily @hariharans29 @tianleiwu Should this merge as is?

**Comment by jeffdaily** [comment]
> @jeffdaily @hariharans29 @tianleiwu Should this merge as is?

I think so.  I just rebased against latest upstream main.  I also ported two cudagraph tests from the shared lib option.  They pass for me when I enabled the shared lib build.

```
Note: Google Test filter = *graph*
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from CApiTest
[ RUN      ] CApiTest.basic_cuda_graph
[       OK ] CApiTest.basic_cuda_graph (956 ms)
[ RUN      ] CApiTest.hip_graph_with_shape_nodes
2024-01-10 21:27:37.244050431 [W:onnxruntime:, inference_session.cc:1754 Initialize] This model has shape massaging nodes that will execute on CPU. Use the CUDA/HIP Graph feature with caution. As long as the intermediate shapes produced in the model using the representative input used to capture the CUDA/HIP graph, will match the shapes produced in the model for other inputs of the same shape as the representative input (common case), it is safe to use the CUDA/HIP Graph feature.
2024-01-10 21:27:37.244161810 [W:onnxruntime:, session_state.cc:1166 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
2024-01-10 21:27:37.244177249 [W:onnxruntime:, session_state.cc:1168 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
[       OK ] CApiTest.hip_graph_with_shape_nodes (24 ms)
[ RUN      ] CApiTest.session_options_graph_optimization_level
[       OK ] CApiTest.session_options_graph_optimization_level (0 ms)
[----------] 3 tests from CApiTest (980 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (980 ms total)
[  PASSED  ] 3 tests.
```

**Comment by tianleiwu** [comment]
@cloudhan, Could you take a look and see whether it is good to merge now?

**Comment by cloudhan** [comment]
It has always been ready, but the required pipeline is preventing the PR from being merged...

**Review by tianleiwu — DISMISSED** [review]

**Comment by tianleiwu** [comment]
Lint / Python format pipeline failed. Need to merge latest change from main, then run `lintrunner -a`.

**Comment by jeffdaily** [comment]
@tianleiwu @cloudhan I have rebased against latest main and ran lintrunner -a.

**Review by cloudhan — APPROVED** [review]

**Pull request merged**